lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -401,6 +457,252 @@ LogicalResult ConcatRefOp::inferReturnTypes(
+        auto members = type.getMembers();
+        auto inputs = getInput();
+        if (inputs.size() != members.size())
+          return failure();
+        for (size_t i = 0; i < members.size(); i++) {
+          auto memberType = cast<UnpackedType>(members[i].type);
+          auto inputType = cast<RefType>(inputs[i].getType()).getNestedType();
+          if (inputType != memberType) {
+            emitOpError("input types must match struct field types and orders");
+            return failure();
+          }
+        }
+        return success();
+      })
+      .Default([this](auto &) {
+        emitOpError("Result type must be StructType or UnpackedStructType");
+        return failure();
+      });
+}
+
+//===----------------------------------------------------------------------===//
+// StructExtractOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult StructExtractOp::verify() {
+  /// checks if the type of the result match field type in this struct
+  return TypeSwitch<Type, LogicalResult>(getInput().getType())
+      .Case<StructType, UnpackedStructType>([this](auto &type) {
+        auto members = type.getMembers();
+        auto filedName = getFieldName();
+        auto resultType = getType();
+        for (const auto &member : members) {
+          if (member.name == filedName && member.type == resultType) {
+            return success();
+          }
+        }
+        emitOpError("result type must match struct field type");
+        return failure();
+      })
+      .Default([this](auto &) {
+        emitOpError("input type must be StructType or UnpackedStructType");
+        return failure();
+      });
+}
+
+bool StructExtractOp::canRewire(const DestructurableMemorySlot &slot,
+                                SmallPtrSetImpl<Attribute> &usedIndices,
+                                SmallVectorImpl<MemorySlot> &mustBeSafelyUsed,
+                                const DataLayout &dataLayout) {
+  return TypeSwitch<Type, bool>(getType())
+      .Case<StructType, UnpackedStructType>(
+          [this, &slot, &usedIndices](auto &type) {
+            if (slot.ptr != getInput())
+              return false;
+            auto index = getFieldNameAttr();
+            if (!index)
+              return false;
+            usedIndices.insert(index);
+            return true;
+          })
+      .Default([](auto) { return false; });
+}
+
+DeletionKind StructExtractOp::rewire(const DestructurableMemorySlot &slot,
+                                     DenseMap<Attribute, MemorySlot> &subslots,
+                                     OpBuilder &builder,
+                                     const DataLayout &dataLayout) {
+  return TypeSwitch<Type, DeletionKind>(getType())
+      .Case<StructType, UnpackedStructType>([this, &subslots](auto &type) {
+        auto index = getFieldNameAttr();
+        if (!index)
+          return DeletionKind::Keep;
```

Comment:
This has already been asserted when checking whether the operation can be rewired, no need to do it twice I think.

Reply:
You are certainlyright. That is useless judgement